adapter/stackdriver/log/log.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2017 the Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"text/template"
+	"time"
+
+	"cloud.google.com/go/logging"
+	xctx "golang.org/x/net/context"
+	"google.golang.org/api/option"
+
+	"istio.io/mixer/adapter/stackdriver/config"
+	"istio.io/mixer/adapter/stackdriver/helper"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/pkg/pool"
+	"istio.io/mixer/template/logentry"
+)
+
+type (
+	makeClientFn func(ctx xctx.Context, projectID string, opts ...option.ClientOption) (*logging.Client, error)
+
+	logFn func(logging.Entry)
+
+	builder struct {
+		makeClient makeClientFn
+		types      map[string]*logentry.Type
+	}
+
+	info struct {
+		labels []string
+		tmpl   *template.Template
+		req    *config.Params_LogInfo_HttpRequestMapping
+	}
+
+	handler struct {
+		now func() time.Time // used for testing
+
+		logger adapter.Logger
+		client io.Closer
+		log    logFn
+		info   map[string]info
+	}
+)
+
+var (
+	_ logentry.HandlerBuilder = &builder{}
```

Comment:
is this strictly necessary?  won't the signature of `NewBuilder()` enforce what we are trying to enforce here? It seems like the `handler` related var is the only additive check.

also, I feel like this stuff could use an explanatory comment.

Reply:
Why? It's a pretty standard practice in go and I don't really think it's worth commenting (we'd say something like "ensure builder implements interface foo" but that's exactly what the code says too). Personally, I write it when I create the struct to ensure I have all the interfaces I want (before I write the actual funcs themselves). Leaving them around shows the intent. Plus, only logentry.HandlerBuilder is enforced, logentry.Handler is not enforced by return values.

Reply:
So, it seems like we agree that `logentry.HandlerBuilder` is enforced. If you want to leave this line for redundancy and consistency, I guess that's OK.  My original point, however, was that this line in particular is not necessary.

IMHO, anything that uses `_` for magic purposes should have a comment (here maybe: `ensure type safety`). This is similar to how I feel about imports with the blank identifier.